src/main/java/com/opentable/components/filterorder/FilterOrderResolver.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.opentable.components.filterorder;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import javax.servlet.Filter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.boot.web.servlet.FilterRegistrationBean;
+
+/**
+ * {@link FilterOrderResolver} resolves the order of servlet filters according to injected {@link OrderDeclaration}s.
+ *
+ * <p>
+ * The total ordering is as follows.
+ * First come the filters for which no order was declared, in original injection order.
+ * Next are the filters for which an order was declared, in the order resolved by this class.
+ * If there is a last filter, it comes last.
+ *
+ * <p>
+ * If more than one last filter is declared, or if the order declarations create a cycle, {@link #resolve} throws
+ * a runtime exception with helpful diagnostic information.
+ *
+ * <p>
+ * Inject this just before you initialize the servlet container. It will call
+ * {@link FilterRegistrationBean#setOrder}} on the beans to enforce their order.
+ *
+ * @see OrderDeclaration
+ */
+public class FilterOrderResolver {
+    private static final Logger LOG = LoggerFactory.getLogger(FilterOrderResolver.class);
+
+    private final List<FilterRegistrationBean<?>> filterRegistrationBeans;
+    private final Collection<OrderDeclaration> orderDeclarations;
+
+    @Inject
+    public FilterOrderResolver(
+            final List<FilterRegistrationBean<?>> filterRegistrationBeans,
+            final Optional<Collection<OrderDeclaration>> orderDeclarations) {
+        this.filterRegistrationBeans = filterRegistrationBeans;
+        this.orderDeclarations = orderDeclarations.orElse(Collections.emptySet());
+    }
+
+    @PostConstruct
+    void resolve() {
+        if (orderDeclarations.isEmpty()) {
+            LOG.debug("no order declarations; will not perform resolution");
+            return;
+        }
+
+        final Map<FilterRegistrationBean<?>, Integer> originalIndexes = new HashMap<>();
```

Comment:
Alternatively:
```
final Map<FilterRegistrationBean<?>, Integer> originalIndexes = IntStream.range(0, filterRegistrationBeans.size())
         .boxed()
         .collect(Collectors.toMap(filterRegistrationBeans::get, Function.identity()));
```

Reply:
Woah!  : ]  That is an alternative indeed.

The number of entries here is usually very small (by default, with `@RestHttpServer`, 4).  Unless performance really is an identifiable problem, prefer simpler constructions that might waste a few resources over more verbose ones.